File: Java/Beginner/k_backtracking/boj_15649.java
package Beginner.k_backtracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class boj_15649 {
    private static final List<Integer> lst = new ArrayList<>();
    private static Boolean[] visited;

    private static void solve(int depth, int n, int m) {
        if (depth == m) {
            lst.forEach(i -> System.out.print(i + 1 + " "));
            System.out.println();
            return;
        }

        for (int i = 0; i < n; i++) {
            if (visited[i]) continue;
            visited[i] = true;
            lst.add(i);
            solve(depth + 1, n, m);
            visited[i] = false;
            lst.remove(lst.size() - 1);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] str = sc.nextLine().split("\s");
        int n = Integer.parseInt(str[0]);
        int m = Integer.parseInt(str[1]);

        visited = new Boolean[n];
        Arrays.fill(visited, false);
        solve(0, n, m);
    }
}

File: Kotlin/src/main/kotlin/beginner/k_backtracking/boj_15649.kt
package beginner.k_backtracking

private lateinit var nums: IntArray
private lateinit var visited: BooleanArray
private val tmp = ArrayDeque<Int>()

private fun dfs(depth: Int, m: Int) {
    if (depth == m) {
        for (i in 0 until tmp.size) print("${tmp[i] + 1} ")
        println()
        return
    }

    for (i in nums.indices) {
        if (visited[i]) continue
        visited[i] = true
        tmp.add(i)
        dfs(depth + 1, m)
        visited[i] = false
        tmp.removeLast()
    }
}

fun boj_15649() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    nums = IntArray(n) { it + 1 }
    visited = BooleanArray(n) { false }

    dfs(0, m)
}